spec/AMQP/20-AMQP_FORMAT.md

```diff
@@ -1,3 +1,118 @@
+
+## `traceparent` AMQP format
+
+The field `traceparent` MUST be encoded and decoded using [binary
+protocol](..\extension-binary.html) and stored as a binary type defined in
+section
+[1.6.19](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-binary)
+of AMQP specification.
+
+Property name MUST be `traceparent` - all lowercase without delimiters.
+
+## `tracestate` AMQP format
+
+The field `tracestate` MUST be encoded and decoded as a string to string map.
+See definition of type map in section
+[1.6.23](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-map)
+of AMQP specification.
+
+---
+
+*NOTE about the strings encoding*
+
+Please note that strings are defined as UTF-8 in AMQP. Should anything be
+different from the HTTP encoded opaque strings? Should we allow to benefit from
+wider character set for better encoding of opaque values? Or this will be too
+error prone?
+
+---
+
+Property name MUST be `tracestate` - all lowercase without delimiters.
+
+## Relations with message id field
+
+The field `message-id` is defined in section
+[3.2.4](http://docs.oasis-open.org/amqp/core/v1.0/amqp-core-messaging-v1.0.html#type-properties)
+of AMQP specification.
+
+Message-id, if set, uniquely identifies a message within the message system. The
+message producer is usually responsible for setting the message-id in such a way
+that it is assured to be globally unique. A broker MAY discard a message as a
+duplicate if the value of the message-id matches that of a previously received
+message sent to the same node.
+
+Trace context identifies the context of a worker who either publish or read the
+message while message-id identifies the content of the message. So the message
+with the same message-id can be send twice with the different values of trace
+context fields in case of retries. Also trace context can be changed from broker
+to broker to identify the broker while message-id will not be changed.
+
+It is recommended to use a unique `parent-id` for every publish of a message. So
+it is not possible that `traceparent` will be identical for the messages with
+the different `message-id`.
+
+## AMQP specific security and privacy considerations
+
+AMQP defines a protocol for a potentially long living messages. Long-term
+storing of `traceparent` and `tracestate` fields may require additional handling
+of security and privacy as that may not be covered by "in-flight" data
+exemptions.
+
+So all the same privacy and security techniques should be applied with the
+potentially more strict requirements.
+
+## Size sensitive environments
+
+TODO: There are many good reason to keep the format and sizes of fields
+unchanged. For the size-sensitive environments, if you ONLY INITIATE the message
+outside of existing context – there may be a possibility to save on size by
+truncating or reusing some fields.
+
+## Batching
+
+TODO: Batching of messages – recommended semantics of TraceContext fields when
```

Comment:
I think we should have a discussion and general guidance for batching, but it probably is related more to the instrumentation and tracing system behavior.

My (very small) experience with AMQP involves Azure ServiceBus and EventHubs, I can miss a lot of details here, but my take on this was:

- SB/EH message could be represented by AMQP message. Or a batch of SB/EH messages is a single AMQP message. If it is the typical case for AMQP, **we should give guidance to the AMQP implementations on how to treat these two cases**.

- every  SB/EH message carry trace context - they probably were produced in different contexts under different spans. Such messages should be annotated during creation, not during publishing. This is instrumentation guidance.

- AMQP message constructed from the batch has new trace context, potentially unrelated to each message in the batch. Relation should be established by something like opencensus links. But again  this is instrumentation and tracing system details, not the protocol.

Reply:
The specification should define a binding to the AMQP "[message format 0](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#definition-MESSAGE-FORMAT)", i.e. the layout that is defined in the [Messaging section of the the AMQP spec](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#section-message-format), but should not go into further detail on HOW those messages are being transferred. 

AMQP is enormously efficient when filling link credit with sending messages in sequence without any special overlaid batching. 

Service Bus and Event Hubs indeed have a special message format that can indeed batch multiple AMQP message format messages into one, but the way to bind to that format is by specifying a binding to the core message format and then have the proprietary batching model pick that up via the binding to the standard message.